File: plugins/extract-django/lifecycle.py
"""The root's states, read off the table the code keeps.

Never off the branches of the methods: the table is the claim and the methods
are held to it. In Django the table is either a `TRANSITIONS` mapping on the
model, beside the `TextChoices` that names the states, or the `@transition`
decorators of django-fsm, which are a table written one edge at a time. An edge
nothing makes, a move into a state the table lacks, and a status assigned
anywhere but a mover are each reported.
"""

from __future__ import annotations

import ast
from dataclasses import dataclass, field as dc_field
from typing import Dict, List, Optional, Tuple

import catalog
from domain import Aggregate, ModelDef
from source import assigned, bases, const_str, dotted, inner_class, keyword, methods


@dataclass
class Move:
    to: str
    on: str
    emits: str
    source: str
    sources: List[str] = dc_field(default_factory=list)  # the states it may be made from, when the decorator says


def choices_of(model: ModelDef, name: str) -> Dict[str, str]:
    """A TextChoices class, as member name to the value stored in the column."""
    node = inner_class(model.node, name)
    if node is None:
        for other in model.module.classes():
            if other.name == name:
                node = other
                break
    if node is None or not any(b.split(".")[-1].endswith("Choices") for b in bases(node)):
        return {}
    out = {}
    for member, value, _ in assigned(node):
        if isinstance(value, ast.Constant) and isinstance(value.value, str):
            out[member] = value.value
        elif isinstance(value, ast.Tuple) and value.elts and isinstance(value.elts[0], ast.Constant):
            out[member] = value.elts[0].value
    return out


def state_of(node: ast.AST, choices: Dict[str, Dict[str, str]]) -> str:
    """`Status.DRAFT` or `"draft"`, either way the value in the column."""
    literal = const_str(node)
    if literal:
        return literal
    name = dotted(node)
    if "." in name:
        holder, member = name.rsplit(".", 1)
        table = choices.get(holder.split(".")[-1], {})
        if member in table:
            return table[member]
    return ""


def status_field(model: ModelDef) -> Optional[str]:
    """The field whose choices are the states. `status` by convention, or the
    one field whose choices name a Choices class."""
    for f in model.fields:
        value = keyword(f.call, "choices")
        if value is not None and dotted(value):
            return f.name
    return "status" if model.field("status") else None


def choice_tables(model: ModelDef) -> Dict[str, Dict[str, str]]:
    out = {}
    for node in list(model.node.body) + list(model.module.tree.body):
        if isinstance(node, ast.ClassDef) and any(b.split(".")[-1].endswith("Choices") for b in bases(node)):
            out[node.name] = choices_of(model, node.name)
    return out


def declared_table(model: ModelDef, choices: Dict[str, Dict[str, str]]) -> Dict[str, List[str]]:
    """`TRANSITIONS = {Status.DRAFT: [Status.ISSUED], ...}`, in the order written."""
    for name, value, _ in assigned(model.node):
        if name != "TRANSITIONS" or not isinstance(value, ast.Dict):
            continue
        table: Dict[str, List[str]] = {}
        for key, targets in zip(value.keys, value.values):
            state = state_of(key, choices)
            if not state:
                continue
            listed = []
            if isinstance(targets, (ast.List, ast.Tuple, ast.Set)):
                for item in targets.elts:
                    to = state_of(item, choices)
                    if to:
                        listed.append(to)
            table[state] = listed
        return table
    return {}


def emits_of(node: ast.AST, events: Dict[str, object]) -> str:
    """The event a mover hands back, as its return annotation names it, by the
    id the catalog knows it under."""
    annotation = getattr(node, "returns", None)
    if annotation is None:
        return ""
    found = ""
    for name in ast.walk(annotation):
        if isinstance(name, ast.Name) and name.id in events:
            found = events[name.id].id
        elif isinstance(name, ast.Constant) and isinstance(name.value, str) and name.value in events:
            found = events[name.value].id
    return found


def movers(model: ModelDef, status: str, choices: Dict[str, Dict[str, str]], events: Dict[str, object]) -> Tuple[List[Move], List[Tuple[str, str]]]:
    """Every method that moves the root: the ones assigning the status field,
    and the ones django-fsm decorates. Also what they assign that no table
    knows, which the caller reports."""
    out: List[Move] = []
    loose: List[Tuple[str, str]] = []
    for node in methods(model.node):
        if node.name.startswith("_"):
            continue
        emits = emits_of(node, events)
        where = model.module.where(node)
        decorator = None
        for dec in getattr(node, "decorator_list", []):
            if dotted(dec).split(".")[-1] == "transition" and isinstance(dec, ast.Call):
                decorator = dec
        if decorator is not None:
            target = state_of(keyword(decorator, "target"), choices)
            source = keyword(decorator, "source")
            froms = []
            if isinstance(source, (ast.List, ast.Tuple, ast.Set)):
                froms = [state_of(item, choices) for item in source.elts]
            elif source is not None:
                one = state_of(source, choices)
                froms = [one] if one else []
            if target:
                out.append(Move(to=target, on=node.name, emits=emits, source=where, sources=[f for f in froms if f]))
            continue
        for assignment in ast.walk(node):
            if not isinstance(assignment, ast.Assign):
                continue
            for target in assignment.targets:
                if not isinstance(target, ast.Attribute) or target.attr != status:
                    continue
                if dotted(target.value) != "self":
                    continue
                to = state_of(assignment.value, choices)
                if to:
                    out.append(Move(to=to, on=node.name, emits=emits, source=model.module.where(assignment)))
                else:
                    loose.append((node.name, model.module.where(assignment)))
    return out, loose


def read(agg: Aggregate, events: Dict[str, object], b) -> Optional[Dict[str, object]]:
    model = agg.root
    status = status_field(model)
    if status is None:
        return None
    choices = choice_tables(model)
    table = declared_table(model, choices)
    moves, loose = movers(model, status, choices, events)

    if not table:
        if not moves:
            return None
        # django-fsm writes the table one edge at a time; a model that only
        # assigns its status has no table at all, and a lifecycle read off the
        # assignments would be the branches talking.
        if not any(m.sources for m in moves):
            b.warn(agg.id, "%s moves its %s but declares no TRANSITIONS: the table is the claim, the methods are held to it" % (model.name, status))
            return None
        table = {}
        for move in moves:
            for source in move.sources:
                table.setdefault(source, [])
                if move.to not in table[source]:
                    table[source].append(move.to)
            table.setdefault(move.to, [])

    states = list(table.keys())
    # The order the Choices class writes them in is the order a reader expects,
    # and the first is where a new root starts.
    declared = next(iter(choices.values()), {})
    if declared:
        ordered = [value for value in declared.values() if value in table or any(value in tos for tos in table.values())]
        for state in states:
            if state not in ordered:
                ordered.append(state)
        states = ordered
    for move in moves:
        if move.to not in states:
            states.append(move.to)
            b.warn(agg.id, "%s moves to %r, which the table does not list" % (move.on, move.to))

    transitions = []
    made = set()
    for move in moves:
        froms = [state for state, tos in table.items() if move.to in tos]
        if move.sources:
            froms = [f for f in move.sources if f in states] or move.sources
        if not froms:
            continue
        for state in froms:
            transitions.append(catalog.transition(state, move.to, move.on, move.emits, move.source))
            made.add((state, move.to))
    for state, tos in table.items():
        for to in tos:
            if (state, to) not in made:
                b.warn(agg.id, "the table has %s -> %s and no method of %s makes it" % (state, to, model.name))
    for name, where in loose:
        b.warn(agg.id, "%s assigns %s to something the states do not name (%s)" % (name, status, where))

    if not transitions:
        return None
    return {"states": states, "transitions": transitions}

File: plugins/extract-java/src/org/portolan/extract/Lifecycle.java
package org.portolan.extract;

import com.sun.source.tree.AssignmentTree;
import com.sun.source.tree.ClassTree;
import com.sun.source.tree.ExpressionTree;
import com.sun.source.tree.MethodInvocationTree;
import com.sun.source.tree.MethodTree;
import com.sun.source.tree.Tree;
import com.sun.source.tree.VariableTree;
import com.sun.source.util.TreeScanner;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * The root's states, read off the table the code keeps.
 *
 * Never off the branches of the methods: the table is the claim and the methods
 * are held to it. In Java the table is a {@code TRANSITIONS} map beside the
 * status enum, and the mover is the method assigning {@code this.status}. An
 * edge nothing makes, a move into a state the table lacks, and a status
 * assigned outside a mover are each reported.
 */
final class Lifecycle {

    private record Move(String to, String on, String emits, String source) {}

    private Lifecycle() {}

    static Map<String, Object> read(Domain.Aggregate aggregate, Map<String, Events.Found> events, Protocol.Builder b) {
        ClassTree statuses = statusEnum(aggregate);
        if (statuses == null) {
            return null;
        }
        Map<String, List<String>> table = table(statuses);
        List<String> states = new ArrayList<>(constants(statuses));
        if (table.isEmpty()) {
            b.warn(aggregate.id(), statuses.getSimpleName() + " names the states but declares no TRANSITIONS: the table is the claim, and the methods are held to it");
            return null;
        }

        List<Move> moves = new ArrayList<>();
        List<String> loose = new ArrayList<>();
        Source.Unit unit = aggregate.rootUnit;
        for (MethodTree method : Source.methods(aggregate.root)) {
            String emits = emits(method, events);
            new TreeScanner<Void, Void>() {
                @Override
                public Void visitAssignment(AssignmentTree node, Void ignored) {
                    String target = node.getVariable().toString();
                    if (target.equals("this.status") || target.equals("status")) {
                        String to = state(node.getExpression());
                        if (to.isEmpty()) {
                            loose.add(method.getName() + " (" + unit.where(node) + ")");
                        } else {
                            moves.add(new Move(to, method.getName().toString(), emits, unit.where(node)));
                        }
                    }
                    return super.visitAssignment(node, ignored);
                }
            }.scan(method, null);
        }

        for (Move move : moves) {
            if (!states.contains(move.to())) {
                states.add(move.to());
                b.warn(aggregate.id(), move.on() + " moves to " + move.to() + ", which the states do not name");
            }
        }

        List<Object> transitions = new ArrayList<>();
        Set<String> made = new LinkedHashSet<>();
        for (Move move : moves) {
            for (Map.Entry<String, List<String>> entry : table.entrySet()) {
                if (!entry.getValue().contains(move.to())) {
                    continue;
                }
                transitions.add(Catalog.transition(entry.getKey(), move.to(), move.on(), move.emits(), move.source()));
                made.add(entry.getKey() + "->" + move.to());
            }
        }
        for (Map.Entry<String, List<String>> entry : table.entrySet()) {
            for (String to : entry.getValue()) {
                if (!made.contains(entry.getKey() + "->" + to)) {
                    b.warn(aggregate.id(), "the table has " + entry.getKey() + " -> " + to + " and no method of " + aggregate.root.getSimpleName() + " makes it");
                }
            }
        }
        for (String where : loose) {
            b.warn(aggregate.id(), where + " assigns the status to something the states do not name");
        }

        if (transitions.isEmpty()) {
            return null;
        }
        return Catalog.map("states", states, "transitions", transitions);
    }

    /** The enum of the aggregate that carries the table, or names the status field's type. */
    private static ClassTree statusEnum(Domain.Aggregate aggregate) {
        String type = "";
        for (VariableTree field : Source.fields(aggregate.root)) {
            if (field.getName().contentEquals("status")) {
                type = Source.simple(field.getType().toString());
            }
        }
        for (Source.Unit unit : new LinkedHashSet<>(aggregate.unitOf.values())) {
            for (ClassTree declared : unit.classes()) {
                if (Source.isEnum(declared) && (declared.getSimpleName().contentEquals(type) || !table(declared).isEmpty())) {
                    return declared;
                }
            }
        }
        return null;
    }

    /** The constants an enum declares, in the order it declares them. */
    private static List<String> constants(ClassTree type) {
        List<String> out = new ArrayList<>();
        for (Tree member : type.getMembers()) {
            if (member instanceof VariableTree field
                    && field.getModifiers().getFlags().contains(javax.lang.model.element.Modifier.STATIC)
                    && field.getInitializer() != null
                    && field.getInitializer().toString().startsWith("/*public static final*/")) {
                out.add(field.getName().toString());
            }
        }
        if (!out.isEmpty()) {
            return out;
        }
        // Older shapes of the tree spell an enum constant as a plain field of
        // the enum's own type.
        for (VariableTree field : Source.constants(type)) {
            if (Source.simple(field.getType().toString()).contentEquals(type.getSimpleName())) {
                out.add(field.getName().toString());
            }
        }
        return out;
    }

    /** `TRANSITIONS = Map.of(PENDING, List.of(AUTHORIZED, …), …)`, in the order written. */
    private static Map<String, List<String>> table(ClassTree type) {
        Map<String, List<String>> out = new LinkedHashMap<>();
        for (VariableTree field : Source.constants(type)) {
            if (!field.getName().contentEquals("TRANSITIONS") || field.getInitializer() == null) {
                continue;
            }
            if (!(field.getInitializer() instanceof MethodInvocationTree call)) {
                continue;
            }
            List<? extends ExpressionTree> args = call.getArguments();
            for (int i = 0; i + 1 < args.size(); i += 2) {
                String from = state(args.get(i));
                List<String> targets = new ArrayList<>();
                if (args.get(i + 1) instanceof MethodInvocationTree list) {
                    for (ExpressionTree target : list.getArguments()) {
                        String to = state(target);
                        if (!to.isEmpty()) {
                            targets.add(to);
                        }
                    }
                }
                if (!from.isEmpty()) {
                    out.put(from, targets);
                }
            }
        }
        return out;
    }

    /** `AUTHORIZED` or `PaymentStatus.AUTHORIZED`, either way the state's name. */
    private static String state(ExpressionTree expression) {
        String text = expression.toString();
        if (text.contains("(") || text.contains("\"")) {
            return "";
        }
        int cut = text.lastIndexOf('.');
        return cut < 0 ? text : text.substring(cut + 1);
    }

    /** The event a mover hands back, as its return type names it. */
    private static String emits(MethodTree method, Map<String, Events.Found> events) {
        if (method.getReturnType() == null) {
            return "";
        }
        Events.Found found = events.get(Source.simple(method.getReturnType().toString()));
        return found == null ? "" : found.id();
    }
}
